### PR TITLE
.Net - Add support for instruction templating on Agents

### PR DESCRIPTION
**Pull request by crickman**
### Motivation and Context

1. Agent instructions support template parameters 
2. Agent yaml identical / interchangeable with prompt config yaml

### Description

Templating agents introduces much more dynamic capabilities, beyond the standard assistant API.

As part of this, the agent template and kernel-function template have converged.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Experimental/Agents/Internal/Agent.cs [hunk 1]
```diff
@@ -181,6 +196,12 @@ private AgentPluginImpl DefinePlugin()
         return new AgentPluginImpl(this, functionAsk);
     }
 
+    private IPromptTemplate DefinePromptTemplate(PromptTemplateConfig config)
+    {
+        var factory = new KernelPromptTemplateFactory(); // $$$ HANDLEBARS
```
Comment (glahaye) [inline dotnet/src/Experimental/Agents/Internal/Agent.cs:201]: Should something happen with this comment?

File: dotnet/samples/KernelSyntaxExamples/Example70_Agents.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ public static async Task RunAsync()
             return;
         }
 
-        // "Hello agent"
+        //// "Hello agent"
```
[the remark this answers is not in the document]
Comment (crickman) [reply dotnet/samples/KernelSyntaxExamples/Example70_Agents.cs:36]: Funny...there is a convention in dotnet that requires double-dashes certain cases when a specific code-analysis rule is enabled...

...but yeah, no...fixing

File: dotnet/src/Experimental/Agents/Internal/Agent.cs (hunk 1 above)
Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/Agent.cs:201]: Yes...this seeped into my awareness and haunted me over the weekend...good call out. 

If I understand correctly, the appropriate template factory needs to be used depending on the template-type (defined withing the template).  This means its possible for the caller to use the wrong factory with a given template.

Ideally, I'd expect a single entry point that was able to fork on whether handlerbars or "kernel prompt"...zooming into thsi now for re-analysis.

Comment (crickman) [reply dotnet/src/Experimental/Agents/Internal/Agent.cs]: Updated

**Review by glahaye — APPROVED** [review]

**Review by alliscode — APPROVED** [review]

**Pull request merged**